### DISCUSSION TITLE
How to use custom SPI pins on ESP32-S3?

### DISCUSSION THREAD

**Discussion by S5NC** [jgromes/RadioLib#1051] [Q&A] [answered]
I tried many things but none work. Documentation:

```cpp
SPIClass spi(VSPI);
SPISettings spiSettings(2000000, MSBFIRST, SPI_MODE0);
SX1262 radio = new Module(cs, irq, rst, gpio, spi, spiSettings);

void setup() {
  spi.begin();
  (...)
  radio.begin();
}
```

I want to use the normal SPI, and keep normal SPI settings. I can leave the spiSettings field empty for it to default to `RADIOLIB_DEFAULT_SPI_SETTINGS`. I don't see where to put my own pins.

My code is 
```cpp
#include <RadioLib.h>
#include <SPI.h>

#define LORA_CS 14
#define LORA_SCK 21
#define LORA_MOSI 38
#define LORA_MISO 39
#define LORA_RESET 40
#define LORA_BUSY 41
#define LORA_DIO1 42

SPI.setPins???(LORA_SCK, LORA_MISO, LORA_MOSI, LORA_SS); // I don't know what to put here

SX1262 radio = new Module(LORA_CS, LORA_DIO1, LORA_RESET, LORA_BUSY);

void setup() {
  Serial.begin(115200);

  // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}
```

**Comment by HeadBoffin** [comment]
> I don't want to use VSPI

Bad news, that's the Arduino default, but you could just blank the V out of your mind and you get SPI.

These should help:

https://espressif-docs.readthedocs-hosted.com/projects/arduino-esp32/en/latest/api/spi.html
and
https://randomnerdtutorials.com/esp32-spi-communication-arduino/

**Comment by S5NC** [comment]
```cpp
#include <RadioLib.h>
#include <SPI.h>

#define LORA_CS 14
#define LORA_SCK 21
#define LORA_MOSI 38
#define LORA_MISO 39
#define LORA_RESET 40
#define LORA_BUSY 41
#define LORA_DIO1 42

#if CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
#define VSPI FSPI
#endif

//SPI.setPins???(LORA_SCK, LORA_MISO, LORA_MOSI, LORA_SS);

// uninitalised pointers to SPI objects
SPIClass * vspi = NULL; // What does this do? Is it needed?

SX1262 radio = new Module(LORA_CS, LORA_DIO1, LORA_RESET, LORA_BUSY, vspi);

void setup() {
  Serial.begin(115200);

  vspi = new SPIClass(VSPI);
  vspi->begin(LORA_SCK, LORA_MISO, LORA_MOSI, LORA_CS);

  // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}
```

```
error: no matching function for call to 'Module(int, int, int, int, SPIClass*&)'
error: invalid conversion from 'SPIClass*' to 'uint32_t' {aka 'unsigned int'} [-fpermissive]
error: invalid user-defined conversion from 'SPIClass*' to 'SPIClass&' [-fpermissive]
error: invalid conversion from 'SPIClass*' to 'uint8_t' {aka 'unsigned char'} [-fpermissive]
error: invalid conversion from 'SPIClass*' to 'uint8_t' {aka 'unsigned char'} [-fpermissive]
error: conversion to non-const reference type 'class SPIClass&' from rvalue of type 'SPIClass' [-fpermissive]
```

**Comment by S5NC** [comment]
```cpp
#include <RadioLib.h>
#include <SPI.h>

#define LORA_CS 14
#define LORA_SCK 21
#define LORA_MOSI 38
#define LORA_MISO 39
#define LORA_RESET 40
#define LORA_BUSY 41
#define LORA_DIO1 42

SPIClass spi(SPI);
SX1262 radio = new Module(LORA_CS, LORA_DIO1, LORA_RESET, LORA_BUSY, spi);

void setup() {
  Serial.begin(115200);

  spi.begin();

  // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}
```

This compiles, but there is nowhere to specify the SPI pins.

**Reply by StevenCellist** [reply]
```spi.begin(SCK, MISO, MOSI, SS);```
from: https://github.com/espressif/arduino-esp32/blob/6c6666adc21ac2862008e06286992b37dc7a858e/libraries/SPI/src/SPI.h#L66

**Reply by S5NC** [reply]
I am trying to uncomment the line with `RADIOLIB_DEBUG_SPI` (https://github.com/jgromes/RadioLib/wiki/Debug-mode) because I get "[SX1262] Initializing ... failed, code -707" but I can't access `BuildOptUser.h`.

**Reply by StevenCellist** [reply]
Then the problem is likely that you made a mistake in one of the pins; triple check them. It depends on your device what pin numbering you should use, famously the RP2040 has two sets of numbering for example.

**Reply by S5NC** [reply]
With `SPIClass loraSPI(SPI);` or `SPIClass loraSPI(HSPI);` it always gives -707 error code, with `SPIClass loraSPI(VSPI);` it doesn't compile because on ESP32-S3 VSPI is not defined, but with `SPIClass loraSPI(FSPI);` it works 5% of the time (95% -707, 5% success)

**Reply by StevenCellist** [reply]
The type of `SPI` _does not (/hardly) matter_! It is all down to your wiring and pin numbering. What board are you using?

**Reply by S5NC** [reply]
It's a [custom design](https://github.com/S5NC/EBYTE_ESP32-S3/blob/main/V1.4%20assembled%20single%20front%20small%20with%20improved%20capacitor.png), but it works perfectly on Meshtastic, no SPI issues whatsoever. The [pins](https://github.com/S5NC/EBYTE_ESP32-S3/blob/main/Schematic%20V1.5%20WIP.pdf) are defined the same.

**Reply by StevenCellist** [reply]
This one's for Mr @jgromes, I am merely an off-the-shelf dude. Maybe something with TCXO or so..

**Reply by S5NC** [reply]
Ah yes I forgot about that! I will pass it into the begin().

**Comment by jgromes** [comment]
@S5NC I suggest slowing down a bit. How did you arrive at the conclusion that the SPI is broken? What is the error code? What is the debug output? If the only error you have seen so far is -707 then that can very well mean the SPI is working and you just have incorrect crystal (as the [Troubleshooting guide](https://github.com/jgromes/RadioLib/wiki/Troubleshooting-Guide) shows. 

Start by observing the symptoms before taking wild guesses ...

**Reply by S5NC** [reply]
I have changed `radio.begin()` to `radio.begin(869.525, 250.0, 11, 5, 12, 22, 8, 1.8, false)`
The Debug output is below:
```
[SX1262] Initializing ... RLB_DBG: RadioLib Debug Info
RLB_DBG: Version:  6.5.0.0
RLB_DBG: Platform: ESP32
RLB_DBG: Compiled: Apr  3 2024 16:32:26

RLB_SPI: CMDW	80	
RLB_SPI: SI		0	
RLB_SPI: SO	AA	AA	
RLB_SPI: CMDW	80	
RLB_SPI: SI		0	
RLB_SPI: SO	A2	A2	
RLB_SPI: CMDR	C0	
RLB_SPI: SI		0	
RLB_SPI: SO	A2	22	
RLB_SPI: CMDR	1D	3	20	
RLB_SPI: SI				0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	
RLB_SPI: SO	A2	A2	A2	A2	53	58	31	32	36	31	20	56	32	44	20	32	44	30	32	0	
RLB_DBG: Found SX126x: RADIOLIB_SX126X_REG_VERSION_STRING:
RLB_DBG: 0000320 53 58 31 32 36 31 20 56 32 44 20 32 44 30 32 00 | SX1261 V2D 2D02.
RLB_DBG: 
RLB_DBG: M	SX126x
RLB_SPI: CMDW	80	
RLB_SPI: SI		0	
RLB_SPI: SO	AA	AA	
RLB_SPI: CMDW	80	
RLB_SPI: SI		0	
RLB_SPI: SO	A2	A2	
RLB_SPI: CMDR	C0	
[...]
RLB_SPI: SO	A2	22	
RLB_SPI: CMDW	8B	
RLB_SPI: SI		B	5	1	0	
RLB_SPI: SO	A2	A2	A2	A2	A2	
RLB_SPI: CMDR	C0	
RLB_SPI: SI		0	
RLB_SPI: SO	A2	22	
RLB_SPI: CMDW	98	
RLB_SPI: SI		D8	DB	
RLB_SPI: SO	A2	A2	A2	
RLB_SPI: CMDR	C0	
RLB_SPI: SI		0	
RLB_SPI: SO	AA	2A	
RLB_SPI: CMDW	86	
RLB_SPI: SI		36	58	66	80	
RLB_SPI: SO	AA	AA	AA	AA	AA	
failed, code -707
```

**Reply by S5NC** [reply]
This time it worked, I changed nothing:
```
[...]
RLB_SPI: SO	A2	A2	A2	A2	18	
RLB_SPI: CMDR	C0	
RLB_SPI: SI		0	
RLB_SPI: SO	A2	22	
RLB_SPI: CMDW	95	
RLB_SPI: SI		4	7	0	1	
RLB_SPI: SO	A2	A2	A2	A2	A2	
RLB_SPI: CMDR	C0	
RLB_SPI: SI		0	
RLB_SPI: SO	A2	22	
RLB_SPI: CMDW	8E	
RLB_SPI: SI		16	4	
RLB_SPI: SO	A2	A2	A2	
RLB_SPI: CMDR	C0	
RLB_SPI: SI		0	
RLB_SPI: SO	A2	22	
RLB_SPI: CMDW	D	8	E7	
RLB_SPI: SI				18	
RLB_SPI: SO	A2	A2	A2	A2	
success!
```
And then I tried to see if it is fixed, I rebooted 10 times and it failed (-707) on each

**Reply by S5NC** [reply]
```cpp
#include <RadioLib.h>
#include <SPI.h>

#define LORA_CS 14
#define LORA_SCK 21
#define LORA_MOSI 38
#define LORA_MISO 39
#define LORA_RESET 40
#define LORA_BUSY 41
#define LORA_DIO1 42

SX1262 radio = new Module(LORA_CS, LORA_DIO1, LORA_RESET, LORA_BUSY);

void setup() {
  Serial.begin(115200);

  SPI.begin(LORA_SCK, LORA_MISO, LORA_MOSI, LORA_CS);

  // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin(869.525, 250.0, 11, 5, 12, 22, 8, 1.8, false);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}
```
and now it's 70% -707, 30% success

**Reply by S5NC** [reply]
Adding `delay(250);` seems to slightly **reduce** the chance of **success**, but this may be random

**Reply by S5NC** [reply]
Still not working, I don't understand the difference between the SPI commands being sent.

**Reply by S5NC** [reply]
[Here](https://github.com/S5NC/EBYTE_ESP32-S3/commit/629b11ad9d7d761fae89d12990412d44fd691827) is the difference between when it works (red) and when it doesn't work (green).
SO output is from [here](https://github.com/jgromes/RadioLib/blob/aa46a0c8b3f4943a6ab8ff90a22f19c25d912e8e/src/Module.cpp#L368), I think it means the buffer going into the ESP32 from the SX1262, but I am not sure.

**Reply by S5NC** [reply]
How can I get debug output of what RadioLib is actually doing?

**Reply by jgromes** [reply]
I like how the first part of my message, which was "slow down a bit", completely passed you ...

The debug output you have showed makes it clear that SPI is working. Now, your error code is -707. Luckily for you, the Troubleshooting guide I linked in my previous post explains what that means - your oscillator setup is invalid. It also tells you what to do to resolve this. Did you follow that guide? If so, report your findings. If not, then feel free to keep trying random guesses.

**Reply by HeadBoffin** [reply]
> How can I get debug output of what RadioLib is actually doing?

Drink less coffee, you've already got the full debug going - that's all those cryptic SPI commands in the full SPI debug output in the "un-reachable" BuildOptUser.h. Which development environment are you using, we may be able to optimise your workflow to help speed things up.

But mostly, as this appears to be a "how do I implement SPI on my very own design that works with Meshtastic", then the short answer is to copy the SPI initialisation code from Meshtastic, the better answer is to create the minimum viable code without any libraries to just say hello to the radio over SPI, the even better answer is to get a very simple SPI sensor and do "SPI hello world" and if that works, hammer it a bit to check that the hardware is reliable.

**Reply by S5NC** [reply]
It fails running [this](https://github.com/jgromes/RadioLib/blob/aa46a0c8b3f4943a6ab8ff90a22f19c25d912e8e/src/modules/SX126x/SX1262.cpp#L20C1-L21C26)
```
  state = setFrequency(freq);
  RADIOLIB_ASSERT(state);
```

**Reply by HeadBoffin** [reply]
In the absence of what value of frequency you set, I'm not sure what we can do to help. Perhaps you could drill down in to that function to set some more debug ...

And just check you aren't asking the module to do something it can't handle?

**Reply by S5NC** [reply]
When setFrequencyRaw returns 911763072 it sometimes doesn't work
When setFrequencyRaw returns 911763072 it sometimes works... Same number.

**Reply by S5NC** [reply]
@HeadBoffin It works sometimes, that's the thing I don't understand

**Reply by HeadBoffin** [reply]
Hence the "try some other SPI module to verify". Because there is no code in RL to run random SPI commands!

**Reply by StevenCellist** [reply]
Dude... **read!!** The solution has been spoiled twice now by @jgromes.

**Reply by S5NC** [reply]
I am using TCXO at 1.8 V, not XTAL

**Reply by S5NC** [reply]
I have found the difference in SPI output to be in the calibration stage.

**Reply by S5NC** [reply]
Is SO or SI sending data from the MCU to the LoRa Radio?

**Reply by S5NC** [reply]
Changing CPU speed from 240 MHz to 80 MHz (like Meshtastic) did not solve the issue.

**Reply by S5NC** [reply]
The difference between AA and A2 in hex is 1 bit. But I don't think it's noise as it happens at the exact same time when it does, on the exact same bit.

**Reply by S5NC** [reply]
```
?010 101?
 0x2 0x5
```
0x2 = STBY_RC
0x5 = failed to execute command

But assertion of state does not fail?

**Reply by jgromes** [reply]
@S5NC I will go through your posts in order.

> I am using TCXO at 1.8 V, not XTAL

What is the source of that information? Ebyte website which can't even spell the word "crystal" correctly? I would have significant doubts if that were the case...

> I have found the difference in SPI output to be in the calibration stage.

A classic symptom of invalid oscillator setup. The SX126x tries to calibrate it's high frequency clock source, and it is unable to do so because the oscillator is not running. On occasion, it can be caused by other things, e.g. frequency being out of range for your module.

> Is SO or SI sending data from the MCU to the LoRa Radio?

Yes, SO is "slave out", SI is "slave in" - same naming convention as all SPI devices.

> Changing CPU speed from 240 MHz to 80 MHz (like Meshtastic) did not solve the issue.

I have no idea why that would do anything.

> The difference between AA and A2 in hex is 1 bit. But I don't think it's noise as it happens at the exact same time when it does, on the exact same bit.

It's quite obviously not noise. It's the SX126x telling you "I cannot complete the command you have requested". The command in this case is 0x98, `RADIOLIB_SX126X_CMD_CALIBRATE_IMAGE`. This is done as part of the frequency configuration. So - does this happen for other frequencies as well, e.g. 868 MHz?

**Reply by S5NC** [reply]
Thank you for your reply, the issue doesn't happen when I set freq to `868` :partying_face:. I am still confused why it sometimes worked and sometimes didn't even when the ESP32-S3 sent the same data via SPI to the SX1262. I will investigate the oscillator issue to see if it is the culprit.

**Reply by jgromes** [reply]
@S5NC not so fast, I don't think we're done here. It seems to be suspiciously similar to https://github.com/jgromes/RadioLib/issues/951 - over there the image calibration was failing at 869.525024 MHz.

**Reply by S5NC** [reply]
Meshtastic calls begin with `freq=869.525024` with no issue on my hardware :smile:

**Reply by S5NC** [reply]
[Here](https://github.com/S5NC/EBYTE_ESP32-S3/commit/ff313b8c795305e3964479e43cd391ee7be7e327) is the difference between the log output of, on the left: `freq=868` successful run (always works), and on the right: `freq=869.525` successful run (random chance of working)

[Here](https://github.com/S5NC/EBYTE_ESP32-S3/commit/aaa2ab3189d67db1b72425a90bb90d80cae7aa24) is the difference between the log output of, on the left: `freq=869.525` successful run (random chance of working), and on the right, the very next run: `freq=869.525` failure (random change of not working)

[Here](https://github.com/S5NC/EBYTE_ESP32-S3/commit/857a188cb28bdce4e1f6395d968133ab08c99c64) is the difference between the log output of, on the left: `freq=869.525` **Meshtastic** successful run (always), and on the right, `freq=869.525` via _**Arduino**_ successful run (random chance of working)

Notice how Meshtastic ignores some -707 and powers through and it all ends up working in the end?

**Reply by S5NC** [reply]
I think the lots of -707 is an anomaly for Meshtastic, usually there are only 2x -707, not 4x. https://github.com/S5NC/EBYTE_ESP32-S3/commit/87c73f91b57ffaad968273d7d327139c6f6e6fe1. Maybe it is also random luck there too.

**Reply by S5NC** [reply]
Tally of 13 tests:
Meshtastic boots with 4x mentions of -707 in the linked custom boot log: 2
Meshtastic boots with 2x mentions of -707 in the linked custom boot log: 11

It seems like the SX1262 works fine in both cases for Meshtastic. There isn't a crash when I try to send messages. I haven't tried with another device to see if those packets are actually transmitted. Maybe the SNR is lower sometimes if the calibration fails?

I wasn't expecting Meshtastic to have this issue, I thought it worked perfectly fine.

**Reply by S5NC** [reply]
I would guess that the frequency calibration must a different start and end value, like D7 and D8, not D8 and D8. Maybe some loose maths is done by the SX1262 to see if the frequency specified (869.525) is within the range, by only doing the sum with the first bytes of a float, but comparing the full memory address. The only reason I could think of which would give a result as observed (sometimes working, sometimes not) would be a random seed generated through environmental noise, but I don't think the SX1262 has any need for a random number, or using memory which isn't flushed between cycles, as running the same thing twice should give the same result on two separate boots, unless timing affects it (moving buffer to memory etc. leading to different data in unused RAM). In this case, having two different starting bytes (D7 vs D8) always ensures that whatever 869.525 is, or that D8 - D7 is positive. Maybe if this wild guess is true, when D8 - D8 is performed, sometimes due to lingering memory in the slots _P, Q, R, S_ of 0xD8_PQ_ - 0x86_RS_ sometimes give negative results and mean the instruction can't be executed (error 5) because the check it performs raises an error. If we set D9, D8 in that order, do we get the same exact error as we are seeing?

**Reply by S5NC** [reply]
> I changed
> ```cpp
> int16_t state = SX126x::calibrateImage(freq - band, freq + band);
> ```
> to
> ```cpp
> int16_t state = SX126x::calibrateImage(freq + band, freq - band);
> ```
> to do this, and...
> 9x -707, 12x success on `freq=869.525`
> 0x -707, 11x successes on `freq=868`
> so it's not the former theory...

^ Ignore this, I forgot to save the modified code! Goes to show how random the success/failure is for 869.525 if it did more successes than failures.

**Reply by S5NC** [reply]
Nevermind it's D8 and DB (letter B not number 8) :rofl:

**Reply by S5NC** [reply]
I think a solution would be to **only** use the sets of calibration values provided in the SX1261/2 datasheet.

![image](https://github.com/jgromes/RadioLib/assets/145265251/d1be5d34-6d1d-41b1-a6de-e106d47251f9)
(From: Page 58 of 114, Semtech, SX1261/2, Data Sheet, DS.SX1261-2.W.APP, Rev. 2.1, December 2021)

When 868 is entered, D7 and D8 are requested, and this always works. If I then switch the frequency to what I actually want using `radio.setFrequency(869.525, false);`, the `false` argument specifying that no calibration should occur for this new frequency, I do not get any -707 errors.

**Reply by jgromes** [reply]
> I think a solution would be to only use the sets of calibration values provided in the SX1261/2 datasheet.

That's the way this was handled before I started messing about with LR11xx driver, which included the calculation that's implemented now. I will revert that change in SX126x - despite testing with couple different SX126x devices, it looks like there are some edge cases in which this calucation doesn't work. I will however keep this calibration available to users, it just won't be used by default.

**Reply by jgromes** [reply]
Pushed in the latest commit, now pre-defined bands will be used preferentially and the calculated coefficients will only be used if your frequency is outside those bands.

**Reply by S5NC** [reply]
> I would guess that the frequency calibration must a different start and end value, like D7 and D8, not D8 and D8. Maybe some loose maths is done by the SX1262 to see if the frequency specified (869.525) is within the range, by only doing the sum with the first bytes of a float, but comparing the full memory address. The only reason I could think of which would give a result as observed (sometimes working, sometimes not) would be a random seed generated through environmental noise, but I don't think the SX1262 has any need for a random number, or using memory which isn't flushed between cycles, as running the same thing twice should give the same result on two separate boots, unless timing affects it (moving buffer to memory etc. leading to different data in unused RAM). In this case, having two different starting bytes (D7 vs D8) always ensures that whatever 869.525 is, or that D8 - D7 is positive. Maybe if this wild guess is true, when D8 - D8 is performed, sometimes due to lingering memory in the slots _P, Q, R, S_ of 0xD8_PQ_ - 0x86_RS_ sometimes give negative results and mean the instruction can't be executed (error 5) because the check it performs raises an error. If we set D9, D8 in that order, do we get the same exact error as we are seeing?

SX1261/2 do have random number generators (RandomNumberGen[0:3], 0x0819:C), so maybe it is a truly random problem.

**Reply by S5NC** [reply]
We could use `GetDeviceErrors` command of opcode `0x17` and check bit 4 of the returned value which is `IMG_CALIB_ERR` to see if the image calibration was the problem.

**Reply by S5NC** [reply]
When the image calibration fails, it always takes slightly longer than when it succeeds.
My testing shows:

(D9, DB) -> stable
(D8, DB) -> unstable
(D7, DB) -> stable

(D9, DA) -> unstable
(D8, DA) -> unstable
(D7, DA) -> unstable

**Comment by HeadBoffin** [comment]
Can you put your 8 layer PCB design up on to your Git repro just in case there is something up with the routing.

**Comment by S5NC** [comment, marked as answer]
Here is the working code which I am using to set custom SPI pins on the ESP32-S3 and SX1262:

```cpp
#include <RadioLib.h>
#include <SPI.h>

#define SX1262_CS 14
#define SX1262_SCK 21
#define SX1262_MOSI 38
#define SX1262_MISO 39
#define SX1262_RESET 40
#define SX1262_BUSY 41
#define SX1262_DIO1 42

#define SX1262_DIO2_AS_RF_SWITCH 1
#define SX1262_TXEN RADIOLIB_NC
#define SX1262_RXEN 10

SX1262 radio = new Module(SX1262_CS, SX1262_DIO1, SX1262_RESET, SX1262_BUSY);

void checkState(int16_t state) {
  if (state != RADIOLIB_ERR_NONE) {
    Serial.print(F("[SX1262] configuration failed with code "));
    Serial.println(state);
    while (true) {}
  }
}

void setup() {

  Serial.begin(115200);

  SPI.begin(SX1262_SCK, SX1262_MISO, SX1262_MOSI, SX1262_CS);
  Serial.printf("SPI.begin(SCK=%d, MISO=%d, MOSI=%d, CS=%d)\n", SX1262_SCK, SX1262_MISO, SX1262_MOSI, SX1262_CS);
  SPI.setFrequency(4000000);

  Serial.println(F("[SX1262] Initialising ... "));

  checkState(radio.begin(868.0, 250.0, 11, 5, 0x2B, 8, 16, 1.8, false));
  checkState(radio.setFrequency(869.525, false)); // you can remove this line and just put 869.525 instead of 868.0 in the line above once a release with the calibration range has been released.
  checkState(radio.setCurrentLimit(140));
  checkState(radio.setCRC(true));
  checkState(radio.setDio2AsRfSwitch(SX1262_DIO2_AS_RF_SWITCH));
  radio.setRfSwitchPins(SX1262_RXEN, SX1262_TXEN); // void
  checkState(radio.setRxBoostedGainMode(true));

  Serial.println(F("[SX1262] All settings succesfully changed!"));

}

void loop() {}
```

**Reply by HeadBoffin** [reply]
Thank you for sharing the solution, much appreciated
